src/module.c

```diff
@@ -2486,6 +2506,342 @@ RedisModuleString *RM_ListPop(RedisModuleKey *key, int where) {
+ * tail of the list. Here, -1 means the last element, -2 means the penultimate
+ * and so forth.
+ *
+ * When no value is found at the given key and index, an NULL is returned and
+ * `errno` is set as follows:
+ *
+ * - EINVAL if key is NULL.
+ * - ENOTSUP if the key is not a list.
+ * - EBADF if the key is not opened for reading.
+ * - EDOM if the index is out of range.
+ */
+RedisModuleString *RM_ListGet(RedisModuleKey *key, long index) {
+    if (!key) {
+        errno = EINVAL;
+        return NULL;
+    } else if (!key->value || key->value->type != OBJ_LIST) {
+        errno = ENOTSUP;
+        return NULL;
+    } else if (!(key->mode & REDISMODULE_READ)) {
+        errno = EBADF;
+        return NULL;
+    }
+    RedisModuleString *value = listTypeIndex(key->value, index);
+    if (value) {
+        autoMemoryAdd(key->ctx, REDISMODULE_AM_STRING, value);
+        return value;
+    } else {
+        errno = EDOM;
+        return NULL;
+    }
+}
+
+/* Replaces the element at index `index` in the list stored at `key`.
+ *
+ * The index is zero-based, so 0 means the first element, 1 the second element
+ * and so on. Negative indices can be used to designate elements starting at the
+ * tail of the list. Here, -1 means the last element, -2 means the penultimate
+ * and so forth.
+ *
+ * On success, REDISMODULE_OK is returned. On failure, REDISOMDULE_ERR is
+ * returned and `errno` is set as follows:
+ *
+ * - EINVAL if key or value is NULL.
+ * - ENOTSUP if the key is not a list.
+ * - EBADF if the key is not opened for writing or if the key has an iterator
+ *   started.
+ * - EDOM if the index is out of range.
+ */
+int RM_ListSet(RedisModuleKey *key, long index, RedisModuleString *value) {
+    if (!key || !value) {
+        errno = EINVAL;
+        return REDISMODULE_ERR;
+    } else if (!key->value || key->value->type != OBJ_LIST) {
+        errno = ENOTSUP;
+        return REDISMODULE_ERR;
+    } else if (!(key->mode & REDISMODULE_WRITE) || key->iter) {
+        errno = EBADF;
+        return REDISMODULE_ERR;
+    }
+    if (listTypeSet(key->value, index, value)) {
+        return REDISMODULE_OK;
+    } else {
+        errno = EDOM;
+        return REDISMODULE_ERR;
+    }
+}
+
+/* Inserts an element before or after the element at the given index.
+ *
+ * The `where` argument is either REDISMODULE_LIST_BEFORE or
+ * REDISMODULE_LIST_AFTER.
+ *
+ * The index is zero-based, so 0 means the first element, 1 the second element
+ * and so on. Negative indices can be used to designate elements starting at the
+ * tail of the list. Here, -1 means the last element, -2 means the penultimate
+ * and so forth.
+ *
+ * On success, REDISMODULE_OK is returned. On failure, REDISOMDULE_ERR is
+ * returned and `errno` is set as follows:
+ *
+ * - EINVAL if key or value is NULL.
+ * - ENOTSUP if the key is not a list.
+ * - EBADF if the key is not opened for writing or if the key has an iterator
+ *   started.
+ * - EDOM if the index is out of range.
```

Comment:
Consider ERANGE which is more obvious?

Reply:
Yes, I've considered ERANGE. `man 3 errno` says

       EDOM            Mathematics  argument  out  of  domain of function (POSIX.1,
                       C99).

       ERANGE          Result too large (POSIX.1, C99).

ERANGE is set by e.g. `scanf`:

       ERANGE The result of an integer conversion would exceed the size that can be
              stored in the corresponding integer type.

My conclusion is that EDOM is about the input to the function while ERANGE is about the result not being representable in the output type (codomain).

Reply:
That's technically correct, I don't recall using `EDOM` so I thought `ERANGE` is more readable and `errno` is anyway taken quite liberally even by libc. Anyway OK either way.

Reply:
OK good. (I used EDOM in the Streams API as well, so if we change this we should change there too.)

What becomes intuition (or easy to remember) is affected by which words are used. Probably I should avoid the word "range" in the description, e.g. change "EDOM if the index is out of range" to "EDOM if the index does not refer to any element in the list" or so.